Show results screen when reaching the end of a failed replay

peppy opened the pull request:
Addresses https://github.com/ppy/osu/discussions/30577.

Consider this a RFC. There's multiple ways this could be implemented, but I went for the most stand-alone approach, bypassing all the score preparation and progress logic (because it's probably simplest this way?).


https://github.com/user-attachments/assets/88d148ab-a37b-4434-bc42-668f41b62dd3

peppy committed:
Show results screen when reaching the end of a failed replay

Addresses https://github.com/ppy/osu/discussions/30577.

Consider this a RFC. There's multiple ways this could be implemented,
but I went for the most stand-alone approach, bypassing all the score
preparation and progress logic (because it's probably simplest this
way?).

bdach commented:
Visually I guess this is fine. The code I have more issues with. The fail flow feels very haphazard. There is now

- `Player.onFail()`
- `Player.Configuration.AllowFailAnimation`
- `Player.ShowFailOverlay`
- `Player.OnFail()`

which are weirdly intermingled and interdependent and three of those are overridable. I guess the diff is relatively slim in isolation so it's difficult for me to attempt to justify some huge refactor but I'd say that this is the last possible change I'd be willing to let slide with these flows, next time these get touched they are 100% getting refactored to something saner.

peppy commented:
I'll give this another attempt while also addressing concerns in the issue (wanting to pause and rewind instead of show results), which basically requires not playing the fail animation anyway (might be able to use that flag).

Gabixel commented:
I think this also covers #22298 (by simply not showing the fail screen at all in replays).

ziv-vy commented:
OH MY GOD MY EARS PEPPY WHAT HAVE YOU DONE
